### PR TITLE
built an api for creating keys,removed changemasterkey API, removed c…

### PR DESCRIPTION
**Pull request by kashishchakarvarti**
…hangeclientkey api

### CODE REVIEW

File: server.js [hunk 1]
```diff
@@ -358,7 +358,7 @@ function ignoreUrl(requestUrl) {
 
     console.log("Adding Ingnore URLS....");
     var ignoreUrl = [ //for the routes to check whether the particular service is active/not
-      "/api/userService", "/api/customService", "/api/roleService", "/api/status", "/file", "/api/createIndex", "/pages", "/status"
+      "/api/userService", "/api/customService", "/api/roleService", "/api/status", "/file", "/api/createIndex", "/pages", "/status","keys"
```
Comment (nawazdhandala) [inline server.js]: Why do we have "keys" here?

File: services/app.js [hunk 2]
```diff
@@ -171,9 +171,22 @@ module.exports = function() {
...
-                        document.keys.js = _generateKey();
-                        document.keys.master = _generateKey();
+                        document.keys ={}
+                        var  masterObj = {};
+                        var  clientObj = {};
+                        var  arr =[];
+
+                        masterObj['key'] = _generateKey();
+                        masterObj['type'] ='master';
+                        masterObj['name'] = 'master';
```
Comment (nawazdhandala) [inline services/app.js]: Name should be "Master" and not master here.

File: services/app.js [hunk 3]
```diff
@@ -171,9 +171,22 @@ module.exports = function() {
...
+                        arr.push(masterObj);
+
+                        clientObj['key'] = _generateKey();
+                        clientObj['type'] ='client';
+                        clientObj['name'] = 'client';
```
Comment (nawazdhandala) [inline services/app.js]: Change this to "Client"

File: services/app.js [hunk 4]
```diff
@@ -210,6 +223,79 @@ module.exports = function() {
...
+
+            var deferred = q.defer();
+            try {                
+
+                var promises = [];
+
+                console.log("Find AppID  exists or not...");
+                var collection =  global.mongoClient.db(global.keys.globalDb).collection("projects");
+                var findAppIdQuery = collection.find({appId:appId});
+                var findNameQuery  = collection.find({"keys.name":name})
```
Comment (nawazdhandala) [inline services/app.js]: Why is this required?

File: services/app.js [hunk 5]
```diff
@@ -210,6 +223,79 @@ module.exports = function() {
             return deferred.promise;
         },
 
+        createAppKeys: function (appId, keyType,name){
+
+            console.log("Create App Masterkey function...");
```
Comment (nawazdhandala) [inline services/app.js]: Check if keyType is master or client, else reject.

File: services/app.js [hunk 6]
```diff
@@ -210,6 +223,79 @@ module.exports = function() {
...
+
+                                        deferred.resolve(project); 
+                                }
+
+                            });  
+                        }
+                    });
+
+                }else{
+                console.log(" AppID not exists");
```
Comment (nawazdhandala) [inline services/app.js]: Indent properly.

File: services/app.js [hunk 7]
```diff
@@ -423,11 +509,15 @@ module.exports = function() {
     			var _self = this;
 
     			_self.getApp(appId).then(function(project){
-    				if(project.keys.master === key){
-    					deferred.resolve(true);
-    				}else{
-    					deferred.resolve(false);
+                    for(i=0;i<project.keys.length;i++)
```
Comment (nawazdhandala) [inline services/app.js]: Indent.

File: services/app.js (hunk 4 above)
Comment (kashishchakarvarti) [inline services/app.js]: to check whether name exist  and app exist

File: services/app.js (hunk 5 above)
Comment (kashishchakarvarti) [inline services/app.js]: ok

File: services/app.js (hunk 2 above)
Comment (kashishchakarvarti) [inline services/app.js]: coverted  name in lowercase in create/keys api , to find the existing name

File: services/app.js (hunk 5 above)
Comment (kashishchakarvarti) [inline services/app.js]: already done that

File: services/app.js (hunk 4 above)
Comment (nawazdhandala) [inline services/app.js]: Why do we check name exist on ALL the apps. It should check name exist on that particular app.

File: services/app.js (hunk 5 above)
Comment (nawazdhandala) [inline services/app.js]: Where?

File: services/app.js (hunk 2 above)
Comment (nawazdhandala) [inline services/app.js]: When we find existing names, Pluck the names out, convert that array to lower case and then check.

File: services/app.js (hunk 5 above)
Comment (kashishchakarvarti) [inline services/app.js]: in APP.js , before passing into this service function

Comment (nawazdhandala) [inline services/app.js]: Lets have that in services instead of the API layer.

File: api/app/App.js [hunk 8]
```diff
@@ -45,7 +45,7 @@ module.exports = function() {
 
 
      //create a new key for an existing app.
-    global.app.put('/app/keys/:appId', function (req, res) {
+    global.app.put('/app/key/:appId', function (req, res) {
```
Comment (nawazdhandala) [inline api/app/App.js]: Remove API from this file.

File: services/app.js [hunk 9]
```diff
@@ -234,33 +234,36 @@ module.exports = function() {
...
 
-                    findNameQuery.toArray(function(err,projects){        
-                        if (err) {
-                            global.winston.log('error',err);
-                            deferred.reject(err);
+                        nameExist = false;  
+
+                        for(i=0;i<projects[0].keys.length;i++)
+                        {
+                            if(projects[0].keys[i].name.toLowerCase() == name)
```
Comment (nawazdhandala) [inline services/app.js]: === instead of == 

and name.toLowerCase() instead of name

File: services/app.js [hunk 10]
```diff
@@ -234,33 +234,36 @@ module.exports = function() {
...
+                            {
+                                nameExist = true
```
Comment (nawazdhandala) [inline services/app.js]: break the loop when name exists. Efficiency.

File: services/app.js [hunk 11]
```diff
@@ -269,20 +272,22 @@ module.exports = function() {
...
-
                                         deferred.resolve(project); 
                                 }
 
-                            });  
+                            }); 
+
+                        } else{
+                            console.log(" Name already exists");
+                            deferred.reject("Name already exists.");
```
Comment (nawazdhandala) [inline services/app.js]: Change to "Key with name 'Name' already exists"

File: services/app.js [hunk 12]
```diff
@@ -269,20 +272,22 @@ module.exports = function() {
...
                         }
-                    });
 
-                }else{
-                console.log(" AppID not exists");
-                 deferred.reject("AppID not exists.");
-                }                                         
+                    } else{
+                         console.log(" AppID not exists");
+                         deferred.reject("AppID not exists.");
```
Comment (nawazdhandala) [inline services/app.js]: Change to "App with App ID 'AppID' does not exists."

File: services/app.js [hunk 13]
```diff
@@ -509,14 +514,16 @@ module.exports = function() {
     			var _self = this;
 
     			_self.getApp(appId).then(function(project){
+
                     for(i=0;i<project.keys.length;i++)
                     {
-        				if(project.keys[i].key === key && project.keys[i].type =='master'){
-        					deferred.resolve(true);
+        				if(project.keys[i].key === key && project.keys[i].type =='master')
```
Comment (nawazdhandala) [inline services/app.js]: Indent properly.

File: services/app.js [hunk 14]
```diff
@@ -540,7 +547,8 @@ module.exports = function() {
                     {
                        // if(project.keys.master === key || project.keys.js === key){
 
-        				if(project.keys[i].key === key){
+        				if(project.keys[i].key === key)
```
Comment (nawazdhandala) [inline services/app.js]: Indent.

File: services/app.js (hunk 9 above)
Comment (kashishchakarvarti) [inline services/app.js]: already done in API but making changes as you said
